enterprise/suppress-user/internal/badgerdb/badgerdb.go

```diff
@@ -0,0 +1,279 @@
+
+	db *badger.DB
+
+	// lock to prevent concurrent access to db during restore
+	restoringLock sync.RWMutex
+	restoring     bool
+	closeOnce     sync.Once
+	closed        chan struct{}
+}
+
+// NewRepository returns a new repository backed by badgerdb.
+func NewRepository(basePath string, log logger.Logger, opts ...Opt) (*Repository, error) {
+	b := &Repository{
+		log:           log,
+		path:          path.Join(basePath, "badgerdbv3"),
+		maxGoroutines: 1,
+		maxSeedWait:   10 * time.Second,
+	}
+	for _, opt := range opts {
+		opt(b)
+	}
+
+	return b, b.start()
+}
+
+// GetToken returns the current token
+func (b *Repository) GetToken() ([]byte, error) {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock() // release the read lock at the end of the operation
+	if b.restoring {
+		return nil, model.ErrRestoring
+	}
+
+	var token []byte
+	err := b.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get([]byte(tokenKey))
+		if err != nil {
+			return err
+		}
+		err = item.Value(func(val []byte) error {
+			token = val
+			return nil
+		})
+		return err
+	})
+	if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+		return nil, err
+	}
+	return token, nil
+}
+
+// Suppress returns true if the given user is suppressed, false otherwise
+func (b *Repository) Suppress(workspaceID, userID, sourceID string) (bool, error) {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	if b.restoring {
+		return false, model.ErrRestoring
+	}
+
+	keyPrefix := fmt.Sprintf("%s:%s:", workspaceID, userID)
+	err := b.db.View(func(txn *badger.Txn) error {
+		_, err := txn.Get([]byte(keyPrefix + model.Wildcard))
+		if err != badger.ErrKeyNotFound {
+			return err
+		}
+		_, err = txn.Get([]byte(keyPrefix + sourceID))
+		return err
+	})
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// Add adds the given suppressions to the repository
+func (b *Repository) Add(suppressions []model.Suppression, token []byte) error {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	if b.restoring {
+		return model.ErrRestoring
+	}
+	if b.db.IsClosed() {
+		return badger.ErrDBClosed
+	}
+	wb := b.db.NewWriteBatch()
+	defer wb.Cancel()
+
+	for i := range suppressions {
+		suppression := suppressions[i]
+		keyPrefix := fmt.Sprintf("%s:%s:", suppression.WorkspaceID, suppression.UserID)
+		var keys []string
+		if len(suppression.SourceIDs) == 0 {
+			keys = []string{keyPrefix + model.Wildcard}
+		} else {
+			keys = make([]string, len(suppression.SourceIDs))
+			for i, sourceID := range suppression.SourceIDs {
+				keys[i] = keyPrefix + sourceID
+			}
+		}
+		for _, key := range keys {
+			var err error
+			if suppression.Canceled {
+				err = wb.Delete([]byte(key))
+			} else {
+				err = wb.Set([]byte(key), []byte(""))
+			}
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	if err := wb.Set([]byte(tokenKey), token); err != nil {
+		return err
+	}
+	err := wb.Flush()
```

Comment:
for the specific method, error contexting happens one level up, by the caller
```
s.log.Errorf("Failed to add %d suppressions to repository: %w", len(suppressions), err)
```
it wouldn't hurt adding another context here though

Reply:
True but you won't be able to tell which part fails, if the flushing, the first set, the second set or the delete. Unless of course the underlying library returns good enough errors.

Reply:
it wouldn't make much of a difference to me if I didn't know where it failed, since it shouldn't really fail if the filesystem is healthy :)
Nevertheless, I am adding context to all errors, as you suggested